scripts/components/search-results.js

```diff
@@ -1,72 +1,142 @@
-  if (!searchResultsContainer) {
-    return;
-  }
 
-  const searchParams = new URLSearchParams(window.location.search);
-  const query = searchParams.get('query');
-  const routedFrom = searchParams.get('source');
+let searchResultsContainer;
 
-  const highlightSearchTerms = !routedFrom;
-  SearchService(query, highlightSearchTerms).then((response) => {
+
+/**
+ * Generator that yields search result pages from SearchService.
+ * @param {String} query         Query string
+ * @param {String} routedFrom    If routed query, the original search terms
+ * @yield {Object} Page as returned by SearchService.
+ */
+const searchPageGenerator = async function* (query, routedFrom) {
+  let nextOffset = null;
+  do {
+    const response = await SearchService({
+      query,
+      offset: nextOffset,
+      // Only request highlighted terms if they correspond to what the user
+      // typed (ie, not routed queries).
+      highlightSearchTerms: !routedFrom,
+      // If first page, fallback to local search after a 3 second timeout
+      searchTimeoutSeconds: nextOffset ? null : 3
+    });
+    nextOffset = response.nextOffset;
     if (response.routeTo) {
       window.location.replace(response.routeTo);
+      return;
     }
     else {
-      renderResults(searchResultsContainer, query, response.results, routedFrom);
+      yield response;
     }
-  });
+  } while (nextOffset);
 };
 
-const renderResults = (searchResultsContainer, query, results, routedFrom) => {
-  searchResultsContainer.innerHTML = '';
-
-  if (routedFrom) {
-    searchResultsContainer.innerHTML += `
-      <h2 class="title">We’re sorry! We found 0 results for “${routedFrom}.”</h2>
-      <h2 class="title">However, we found results for the related term “${query}.”</h2>
-    `;
+export const initSearchResults = () => {
+  searchResultsContainer = document.getElementById(SEARCH_RESULTS_ID);
+  if (!searchResultsContainer) {
+    return;
   }
 
-  if (!results.length) {
-    searchResultsContainer.innerHTML += `<h2 class="title">No results found</h2>`;
-  }
-  else {
-    searchResultsContainer.innerHTML += `
-      <ol>
-        ${results.map(renderResults_result(routedFrom)).join('')}
-      </ol>
-    `;
-  }
+  const searchParams = new URLSearchParams(window.location.search);
+  const query = searchParams.get('query');
+  const routedFrom = searchParams.get('source');
+
+  let resultsPages = [];
+
+  const pages = searchPageGenerator(query, routedFrom);
+
+  const renderNextPage = async () => {
+    resultsPages.push((await pages.next()).value);
+    render(searchTemplate({
+      resultsPages,
+      query,
+      routedFrom,
+      renderNextPage
+    }), searchResultsContainer);
+  };
+  renderNextPage();
 };
 
+const searchTemplate = ({
+  resultsPages,
+  query,
+  routedFrom,
+  renderNextPage
+}) => {
+  const lastPage = resultsPages[resultsPages.length - 1];
+  const resultsCount = lastPage.resultsCount;
+  const nextOffset = lastPage.nextOffset;
+  return html`
```

Comment:
Is this some lithtml magic?

Reply:
Yes, `html` is the function that wires up DOM interactions with the template string.